### PR TITLE
Support for setting cipher list with bytes

### PR DESCRIPTION
**Pull request by SKlimaRA**
# Description

Set cipher suites using their 2byte codes instead of string names.

Adds two new API's:
```
/* supports 2 byte code from cipher_name[] */
WOLFSSL_API int  wolfSSL_CTX_set_cipher_list_bytes(WOLFSSL_CTX* ctx, const byte* list, const int listSz);
WOLFSSL_API int  wolfSSL_set_cipher_list_bytes(WOLFSSL* ssl, const byte* list, const int listSz);
```

Fixes ZD14669

# Testing

How did you test?

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by dgarske** [comment]
OK to test

File: src/internal.c [hunk 1]
```diff
@@ -24086,6 +24086,114 @@ int SetCipherList(WOLFSSL_CTX* ctx, Suites* suites, const char* list)
...
+    int haveDilithiumSig = 0;
+    int haveAnon         = 0;
+
+
+    if (suites == NULL || list == NULL) {
+        WOLFSSL_MSG("SetCipherListFromBytes parameter error");
+        return 0;
+    }
+
+    for (int i = 0; (i + 1) < listSz; i += 2) {
```
Comment (dgarske) [inline src/internal.c:24108]: `int i` must be declared at top of function.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — CHANGES_REQUESTED** [review]
Please add a test case. Thanks

File: src/ssl.c [hunk 2]
```diff
@@ -11683,6 +11708,29 @@ int wolfSSL_set_cipher_list(WOLFSSL* ssl, const char* list)
...
+    WOLFSSL_ENTER("wolfSSL_set_cipher_list_bytes");
+#ifdef SINGLE_THREADED
+    if (ssl->ctx->suites == ssl->suites) {
+        ssl->suites = (Suites*)XMALLOC(sizeof(Suites), ssl->heap,
+                                       DYNAMIC_TYPE_SUITES);
+        if (ssl->suites == NULL) {
+            WOLFSSL_MSG("Suites Memory error");
+            return MEMORY_E;
+        }
+        *ssl->suites = *ssl->ctx->suites;
```
Comment (dgarske) [inline src/ssl.c:11724]: Please replace this with an `XMEMCPY` -> `XMEMCPY(ssl->suites, ssl->ctx->suites, sizeof(Suites));`.

File: tests/api.c [hunk 3]
```diff
@@ -839,6 +839,173 @@ static int test_for_double_Free(void)
...
+
+#ifndef NO_WOLFSSL_SERVER
+    ctx = wolfSSL_CTX_new(wolfSSLv23_server_method());
+    AssertNotNull(ctx);
+#else
+    ctx = wolfSSL_CTX_new(wolfSSLv23_client_method());
+    AssertNotNull(ctx);
+#endif
+
+    AssertTrue(wolfSSL_CTX_set_cipher_list_bytes(ctx, &cipherList[0U],
```
Comment (dgarske) [inline tests/api.c:998]: Can you also test `wolfSSL_set_cipher_list_bytes`?

**Review by dgarske — CHANGES_REQUESTED** [review]
Great work, just a few minor items to review/fix.

File: src/ssl.c [hunk 4]
```diff
@@ -11683,6 +11708,29 @@ int wolfSSL_set_cipher_list(WOLFSSL* ssl, const char* list)
 #endif
 }
 
+
+int wolfSSL_set_cipher_list_bytes(WOLFSSL* ssl, const byte* list,
+                                  const int listSz)
+{
+    WOLFSSL_ENTER("wolfSSL_set_cipher_list_bytes");
+#ifdef SINGLE_THREADED
+    if (ssl->ctx->suites == ssl->suites) {
```
[the remark this answers is not in the document]
Comment (SKlimaRA) [reply src/ssl.c:11717]: I don't know these top API functions are copy pastes of the string API and only difference is calling different function at the end... so if this is an issue, then it's also in current API

File: tests/api.c (hunk 3 above)
Comment (SKlimaRA) [reply tests/api.c:998]: will do

**Comment by dgarske** [comment]
@SKlimaRA , what use case are you using that needs to use the cipher suite bytes directly and what build options are you using to compile wolfSSL? I am considering having these new API's only accessible with `OPENSSL_EXTRA`. Curious if that will work for your use-case.

**Comment by dgarske** [comment]
> @SKlimaRA , what use case are you using that needs to use the cipher suite bytes directly and what build options are you using to compile wolfSSL? I am considering having these new API's only accessible with `OPENSSL_EXTRA`. Curious if that will work for your use-case.

@SKlimaRA If you are okay with this being wrapped with OPENSSL_EXTRA let me know. I already made the change and the state of the PR looks good.

**Review by dgarske — APPROVED** [review]

**Comment by SKlimaRA** [comment]
@dgarske
Hi, thx for the fixes.
On our end we keep the cipher suites as #define CIPHER_SUITE_NAME 0xTwoByteValue. I don't know who or why decided that but for long time we were patching functions similar to the functions in this PR into the code so we could set the cipher suites using the codes. I guess 2 byte values take less memory than strings.
We are currently doing changes to our wolfSSL configuration and after that work is done we will be able to provide you with some of our configurations through the support. If you would want current temporary config I can provide it via the zendesk thread. Unfortunately we can't use OPENSSL_EXTRA.

**Comment by SKlimaRA** [comment]
@dgarske If you want to hide this new API behind a flag please introduce a new one. So we can enable it without OPENSSL_EXTRA and you can enable the new flag if OPENSSL_EXTRA is defined.

**Comment by dgarske** [comment]
> @dgarske If you want to hide this new API behind a flag please introduce a new one. So we can enable it without OPENSSL_EXTRA and you can enable the new flag if OPENSSL_EXTRA is defined.

Added `WOLFSSL_SET_CIPHER_BYTES` to allow use of the set cipher suite with bytes.

**Review by dgarske — APPROVED** [review]

**Comment by SKlimaRA** [comment]
@dgarske PR looks good, looking forward to the merge.

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: src/ssl.c [hunk 5]
```diff
@@ -11671,17 +11701,49 @@ int wolfSSL_set_cipher_list(WOLFSSL* ssl, const char* list)
             WOLFSSL_MSG("Suites Memory error");
             return MEMORY_E;
         }
-        *ssl->suites = *ssl->ctx->suites;
+        XMEMCPY(ssl->suites, ssl->ctx->suites, sizeof(Suites));
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/ssl.c:11704]: @JacobBarthelmeh I made that change. I'll revert it for @SKlimaRA

[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply src/ssl.c]: Will follow up with an additional pull request if needed outside of this one. The issue noted only affects OPENSSL_ALL builds and could potentially use some changes outside of the code in just this pull request.

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**